### PR TITLE
[SYCL][Group algorithms] Add group sorting algorithms implementation

### PR DESCRIPTION
**Pull request by capatober**
This PR introduce the implementation for the [Group Sort extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc).

The PR includes:
* Feature macro
* `joint_sort` and `sort_over_group` functions
* `default_sorter`
* `group_with_scratchpad`

Algorithms are quite general. It should work with custom data types, custom comparators, custom sorters.

The PR doesn't include:
* `radix_sorter`
* optimized specialization for arithmetic types

Tests are here: https://github.com/intel/llvm-test-suite/pull/438

Signed-off-by: Fedorov, Andrey <andrey.fedorov@intel.com>

### CODE REVIEW

**Comment by bader** [comment]
@capatober, please, create draft pull request for "work in progress".

**Comment by capatober** [comment]
The PR is ready for review.
`clang-format` check differs from `clang-format` from the repo. Seems it should be improved on infrastructure side.

**Comment by capatober** [comment]
@intel/llvm-reviewers-runtime, just kind reminder. Patch is ready for review

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+#ifdef __SYCL_DEVICE_ONLY__
+#include "group_sort_util.hpp"
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+// ---- merge sort implementation
+
+// following to functions could be useless if std::[lower|upper]_bound worked
```
Comment (Pennycook) [inline sycl/include/CL/sycl/detail/group_sort_impl.hpp:24]:
```suggestion
// following two functions could be useless if std::[lower|upper]_bound worked
```

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+// well
+template <typename Acc, typename Value, typename Compare>
+std::size_t my_lower_bound(Acc acc, std::size_t first, std::size_t last,
```
Comment (Pennycook) [inline sycl/include/CL/sycl/detail/group_sort_impl.hpp:27]: Is there any reason to prefix these with `my_` if they're already in `sycl::detail::`?

Same question for `my_upper_bound`, `my_swap` and so on.

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+template <typename Group, typename Iter, typename Compare>
+void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
+                std::uint8_t *scratch) {
+  using T = typename std::iterator_traits<Iter>::value_type;
+  auto id = cl::sycl::detail::Builder::getNDItem<Group::dimensions>();
+  const std::size_t idx = id.get_local_id();
+  const std::size_t local = group.get_local_range().size();
+  const std::size_t chunk = (n - 1) / local + 1;
+
+  // we need to sort within work item firstly
```
Comment (Pennycook) [inline sycl/include/CL/sycl/detail/group_sort_impl.hpp:185]:
```suggestion
  // we need to sort within work item first
```

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  bubble_sort(first, idx * chunk, sycl::min((idx + 1) * chunk, n), comp);
+  id.barrier();
+
+  T *temp = reinterpret_cast<T *>(scratch);
```
Comment (Pennycook) [inline sycl/include/CL/sycl/detail/group_sort_impl.hpp:189]: Is this always guaranteed to be safe?  Should we replace `std::uint8_t` above with `char` or `byte`?

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      : comp(comp_), scratch(scratch_.data()), scratch_size(scratch_.size()) {}
+
+  template <typename Group, typename Ptr>
+  void operator()(Group g, Ptr begin, Ptr end) {
+#ifdef __SYCL_DEVICE_ONLY__
+    using T = typename std::iterator_traits<Ptr>::value_type;
+    if (scratch_size >= memory_required<T>(Group::fence_scope, end - begin))
+      cl::sycl::detail::merge_sort(g, begin, end - begin, comp, scratch);
+      // TODO: it's better to add else branch
+#endif
```
Comment (Pennycook) [inline sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:50]: Should there be something like an `#else` here with a runtime error?

I'm always worried when I see a `__SYCL_DEVICE_ONLY__` with no host code.  The compiler can be quite aggressive in optimizing things out, and it might reach some weird conclusions about whether `begin` and `end` are needed.

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp (hunk 2 above)
Comment (capatober) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:27]: No reason. Just to not mix with `std::` functions

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:27]: For such cases we simply use namespaces like `detail::upper_bound` or `std::upper_bound`.

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+      }
+    }
+  }
+}
+
+template <typename Group, typename Iter, typename Compare>
+void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
+                std::uint8_t *scratch) {
+  using T = typename std::iterator_traits<Iter>::value_type;
+  auto id = cl::sycl::detail::Builder::getNDItem<Group::dimensions>();
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/group_sort_impl.hpp:180]:
```suggestion
  auto id = sycl::detail::Builder::getNDItem<Group::dimensions>();
```

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp (hunk 4 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:189]: While it is exceptionally hard in 2021 to come across the system, that thinks that byte is anything but 8 bits, I agree with John, that the standard only guarantees this to be safe for `unsigned char`, `char`, and `byte`. And yet, I would still prefer this last parameter to be replaced with `sycl::span` or something like that to ensure that we do not get out of bounds of this array. (The last statement probably requires some spec work).

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+      : g(g_), scratch(scratch_) {}
+  Group get_group() const { return g; }
+  sycl::span<std::uint8_t, Extent> get_memory() const { return scratch; }
+};
+
+// ---- sorters
+template <typename Compare = std::less<>> class default_sorter {
+  Compare comp;
+  std::uint8_t *scratch;
+  std::size_t scratch_size;
```
Comment (alexbatashev) [inline sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:35]: This is very similar to my previous comment. A pair of a pointer to an array and its size is either range or span. Since we have span backported to SYCL 2020, can we use it here?

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  default_sorter(sycl::span<std::uint8_t, Extent> scratch_,
+                 Compare comp_ = Compare())
+      : comp(comp_), scratch(scratch_.data()), scratch_size(scratch_.size()) {}
+
+  template <typename Group, typename Ptr>
+  void operator()(Group g, Ptr begin, Ptr end) {
+#ifdef __SYCL_DEVICE_ONLY__
+    using T = typename std::iterator_traits<Ptr>::value_type;
+    if (scratch_size >= memory_required<T>(Group::fence_scope, end - begin))
+      cl::sycl::detail::merge_sort(g, begin, end - begin, comp, scratch);
```
Comment (alexbatashev) [inline sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:48]:
```suggestion
      sycl::detail::merge_sort(g, begin, end - begin, comp, scratch);
```

File: sycl/include/sycl/ext/oneapi/group_sort.hpp [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+// ---- traits
+template <typename T, typename = void> struct has_difference_type {};
+
+template <typename T>
+struct has_difference_type<T,
+#if __cplusplus < 201703
+                           cl::sycl::detail::void_type<
+#else
+                           std::void_t<
+#endif
```
Comment (alexbatashev) [inline sycl/include/sycl/ext/oneapi/group_sort.hpp:37]: Nit. Did clang-format do that? Can we have some alias to avoid it?

File: sycl/include/sycl/ext/oneapi/group_sort.hpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+// ---- sort_over_group
+template <typename Group, typename T, typename Sorter>
+typename std::enable_if<detail::is_sorter<Sorter, Group, T>::value, T>::type
+sort_over_group(Group group, T value, Sorter sorter) {
+#ifdef __SYCL_DEVICE_ONLY__
+  return sorter(group, value);
+#else
+  (void)group;
+  throw runtime_error("Group algorithms are not supported on host device.",
+                      PI_INVALID_DEVICE);
```
Comment (alexbatashev) [inline sycl/include/sycl/ext/oneapi/group_sort.hpp:92]: Can we use SYCL 2020-style exceptions here? AFAIK, they've been implemented recently.

File: sycl/include/sycl/ext/oneapi/group_sort.hpp [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+#endif
+}
+
+template <typename Group, typename T, typename Compare, std::size_t Extent>
+typename std::enable_if<!detail::is_sorter<Compare, Group, T>::value, T>::type
+sort_over_group(
+    cl::sycl::ext::oneapi::experimental::group_with_scratchpad<Group, Extent>
```
Comment (alexbatashev) [inline sycl/include/sycl/ext/oneapi/group_sort.hpp:99]: I think it will still compile, while being much easier to read.

```suggestion
    group_with_scratchpad<Group, Extent>
```

File: sycl/include/sycl/ext/oneapi/group_sort.hpp (hunk 9 above)
Comment (capatober) [reply sycl/include/sycl/ext/oneapi/group_sort.hpp:37]: Since clang-format is passed, yep, this is its work. I've found this one in the repo: https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/stl_type_traits.hpp#L42.

I'm just curious. What does "Nit" mean? I couldn't quickly find what is it.

File: sycl/include/sycl/ext/oneapi/group_sort.hpp (hunk 10 above)
Comment (capatober) [reply sycl/include/sycl/ext/oneapi/group_sort.hpp:92]: Could you please point on the implementation?

File: sycl/include/sycl/ext/oneapi/group_sort.hpp (hunk 11 above)
Comment (capatober) [reply sycl/include/sycl/ext/oneapi/group_sort.hpp:99]: Are not fully qualified names preferred in the repo? If no, I can make it shorter. I'm just asking about conventions.

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp (hunk 4 above)
Comment (capatober) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:189]: I'll think about it. Thanks for the comment

Comment (capatober) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:189]: > Should we replace std::uint8_t above with char or byte?

If so, it's better to make changes for the Spec (https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc).
What do you think? What is a better way to reflect it in the Spec?

Comment (capatober) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:189]: >the standard only guarantees this to be safe for unsigned char, char, and byte

Could you please point on the C++ standard chapter, where it's mentioned?

File: sycl/include/sycl/ext/oneapi/group_sort.hpp (hunk 9 above)
Comment (alexbatashev) [reply sycl/include/sycl/ext/oneapi/group_sort.hpp:37]: Nit means nitpick

File: sycl/include/sycl/ext/oneapi/group_sort.hpp (hunk 10 above)
Comment (alexbatashev) [reply sycl/include/sycl/ext/oneapi/group_sort.hpp:81]: Impl: https://github.com/intel/llvm/pull/4072
Spec: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:exception.class

File: sycl/include/sycl/ext/oneapi/group_sort.hpp (hunk 11 above)
Comment (alexbatashev) [reply sycl/include/sycl/ext/oneapi/group_sort.hpp]: Generally, we use fully qualified names when there can be discrepancies, like when SYCL defines the same type name as STL does. In this case `group_with_scratchpad` is unique to SYCL, so there's no need to duplicate namespaces.

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp (hunk 4 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:197]: I think this is the relevant part of the standard: https://eel.is/c++draft/basic.life#1.1

TL;DR casting and then dereferencing a pointer is only valid if you guarantee proper object size, layout, and alignment. So (let's forget about inheritance for now), in general you can only safely cast a pointer to something, that has the size of 1 byte (and treat it as byte array), and then back to the same type. Now, the byte size is the tricky part. Although modern systems agree that 1 byte == 8 bits, that hasn't always been the case (an interesting [wiki article](https://en.wikipedia.org/wiki/Byte) on that topic). So, from language point of view, one can not assume that `sizeof(uint8_t) == 1`. But `unsigned char`, `char` and `std::byte` are guaranteed to have size of 1 byte. C++ 20 brings some confusion by declaring `sizeof(char8_t) == 1`, but SYCL 2020 is based on C++ 17. This is how I interpret the standard, I may be wrong about that. Hopefully @Pennycook can correct me.

Comment (Pennycook) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:197]: I agree with everything @alexbatashev said.  @capatober, I think the easiest fix is to replace the `uint8_t` in the spec with a `std::byte` -- everything else is fine.  Sorry for not picking up on this during review of the spec; things didn't click until I saw the scratchpad being reinterpreted.

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:197]: @Pennycook and what do you think of embracing `sycl::span` instead of raw pointers? I think, it can help with safety, given that users will be forced to provide size for each range.

Comment (Pennycook) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:197]: I'm concerned that the `sycl::span` variant would be harder to use in the case where the same scratchpad is used for multiple sorts (and, eventually, multiple algorithms).

With a raw pointer / span of bytes, the following usage is allowed (where `array1` and `array2` have different types).
```c++
auto exec = group_with_scratchpad(id.get_group(), span(&scratch[0], local_memory_size));
joint_sort(exec, array1, array1 + N);
joint_sort(exec, array2, array2 + N);
```

The implementation can use `placement new` to allocate objects inside of the `scratch` memory and everything is fine.  Additionally, the implementation can allocate anything it wants in that scratchpad, not just objects of the type being sorted.

If the user has to pass a `span` associated with a particular type, wouldn't they have to handle correct setup and teardown of the memory either side of calling `joint_sort`?  Presumably, the implementation would also lose the freedom to use the memory for anything except objects of the type declared in the `span`.

Comment (v-klochkov) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:197]: Just my opinion: if 8-bits not equal to 1 byte, then many places where SYCL would also not work properly and would never reached the point that is supposed to start some sorting algorithm ))
For example, take a look at this code: https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/aliases.hpp#L85

Comment (alexbatashev) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:197]: @v-klochkov this is OpenCL C type. It is specified to be 8 bits long, rather than 1 byte. See section 6.3.1 of OpenCL C spec.

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+  template <typename T>
+  static constexpr std::size_t memory_required(sycl::memory_scope scope,
+                                               std::size_t range_size) {
+    return range_size * sizeof(T);
+  }
+
+  template <typename T, int dim = 1>
+  static constexpr std::size_t memory_required(sycl::memory_scope scope,
+                                               sycl::range<dim> r) {
+    return 2 * r.size() * sizeof(T);
```
Comment (cperkinsintel) [inline sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:92]: why the `* 2`  ?  Shouldn't `r.size() * sizeof(T)` be sufficient ?  And, if not, why does the `memory_required` variant that takes a `size_t` not also multiply by 2 ?

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+}
+
+template <typename Group, typename Iter, typename Compare>
+void merge_sort(Group group, Iter first, const std::size_t n, Compare comp,
+                std::uint8_t *scratch) {
+  using T = typename GetValueType<Iter>::type;
+  auto id = sycl::detail::Builder::getNDItem<Group::dimensions>();
+  const std::size_t idx = id.get_local_id();
+  const std::size_t local = group.get_local_range().size();
+  const std::size_t chunk = (n - 1) / local + 1;
```
Comment (cperkinsintel) [inline sycl/include/CL/sycl/detail/group_sort_impl.hpp:191]: I'm sorry to be dense. `n` is a size in bytes, correct? So shouldn't `local + 1` be multiplied by the `sizeof(T)` ? 
Maybe I'm misunderstanding `n`, in which case a comment might assist.

**Review by cperkinsintel — COMMENTED** [review]
Does this extension have a doc (under `llvm/sycl/doc/extensions`) ?

**Comment by cperkinsintel** [comment]
Also, we need some tests.

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp (hunk 12 above)
Comment (capatober) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:92]: Sure. It's mentioned in the description of the PR. (https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc)

Regarding your comment about tests. It's also mentioned in the description :) See here: https://github.com/intel/llvm-test-suite/pull/438

**Comment by capatober** [comment]
>Does this extension have a doc (under llvm/sycl/doc/extensions) ?

Sure. It's mentioned in the description of the PR. (https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc)

> Also, we need some tests.

It's also mentioned in the description :) See here: intel/llvm-test-suite#438

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp (hunk 13 above)
Comment (capatober) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:191]: `n` is a difference between 2 iterators that are passed to `joint_sort`. I'd replace `n` with `last` for the parameters list. What do you think? Would it improve readability?

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp (hunk 4 above)
Comment (capatober) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:197]: Quite interesting discussion. I remember words in the C++ standard about guaranteed object size, but I couldn't realize that 8 bits != 1 byte :)

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp (hunk 7 above)
Comment (capatober) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:35]: We consider this option previously during the API discussion. The thing is that if we replaced this with sycl::span, we would need an additional template parameter for sorter (extent). However, extent is not relevant to the `default_sorter`, but to the parameter of its ctor.

File: sycl/include/CL/sycl/detail/group_sort_impl.hpp [hunk 14]
```diff
@@ -98,7 +109,9 @@ void merge(const std::size_t offset, InAcc &in_acc1, OutAcc &out_acc1,
     const std::size_t l_shift_1 = local_start_1 - start_1;
     const std::size_t l_shift_2 = l_search_bound_2 - start_2;
 
-    out_acc1[start_out + l_shift_1 + l_shift_2] = local_l_item_1;
+    // out_acc1[start_out + l_shift_1 + l_shift_2] = local_l_item_1;
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/group_sort_impl.hpp:112]: @capatober will you remove commented code?

Comment (capatober) [reply sycl/include/CL/sycl/detail/group_sort_impl.hpp:112]: Yes. Sorry

**Review by alexbatashev — APPROVED** [review]
LGTM

**Comment by capatober** [comment]
@intel/llvm-reviewers-runtime, is anything else needed to merge the code to the repo? I don't have enough rights for this action

**Comment by alexbatashev** [comment]
> @intel/llvm-reviewers-runtime, is anything else needed to merge the code to the repo? I don't have enough rights for this action

At least you need to make sure CI is green. Also approve from @intel/dpcpp-specification-reviewers is required.

**Comment by capatober** [comment]
@intel/llvm-reviewers-runtime, CI is green. Feel free to approve and merge. Changes for spec are obvious and doesn't require to be approved additionally

**Review by alexbatashev — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        "default_sorter constructor is not supported on host device.");
+#endif
+  }
+
+  template <typename Group, typename T> T operator()(Group g, T val) {
+#ifdef __SYCL_DEVICE_ONLY__
+    auto range_size = g.get_local_range().size();
+    if (scratch_size >= memory_required<T>(Group::fence_scope, range_size)) {
+      auto id = sycl::detail::Builder::getNDItem<Group::dimensions>();
+      uint32_t local_id = id.get_local_id();
```
Comment (Pennycook) [inline sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:66]:
```suggestion
      size_t local_id = id.get_local_linear_id();
```
I've just spotted two things here:
1. I think this should be `size_t` rather than `uint32_t`, because SYCL says IDs are `size_t`
2. Although you're galling `getNDItem<Group::dimensions>()`, doesn't the conversion here assume a 1D work-group?

Comment (Pennycook) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:66]: Do we have any tests for multi-dimensional groups?

**Review by Pennycook — CHANGES_REQUESTED** [review]

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp (hunk 15 above)
Comment (capatober) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:66]: Good catch. Thanks a lot! I was going to add a test case for multi-dimensional groups, but forgot due to a lot of other test cases.

**Review by alexbatashev — APPROVED** [review]
Still LGTM

**Review by Pennycook — APPROVED** [review]
